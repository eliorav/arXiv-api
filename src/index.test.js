const {PREFIXES, SORT_BY, SORT_ORDER} = require('./constants');

const mockResponse = {
	feed: {
		link: [{$: {href: 'LINK'}}],
		updated: ['2021-05-16T00:00:00-04:00'],
		'opensearch:totalResults': [{_: '2'}],
		'opensearch:startIndex': [{_: '0'}],
		'opensearch:itemsPerPage': [{_: '1'}],
		entry: [
			{
				id: ['PAPER_ID'],
				updated: ['UPDATED'],
				published: ['PUBLISHED'],
				title: ['TITLE'],
				summary: ['SUMMARY'],
				author: [{name: ['AUTHOR1']}, {name: ['AUTHOR2']}],
				link: [{$: {title: 'TITLE', href: 'URL', rel: 'REL', type: 'TYPE'}}],
				category: [{$: 'CATEGORY'}],
			},
		],
	},
};

const mockAxiosGet = jest.fn(() => Promise.resolve({data: 'XML'}));
const mockXmlPromisify = jest.fn(() => Promise.resolve(mockResponse));

jest.mock('axios', () => ({
	get: mockAxiosGet,
}));
jest.mock('xml2js');
jest.mock('util', () => ({
	promisify: jest.fn(() => mockXmlPromisify),
}));

const {search, searchWithMeta, parseResponseData} = require('./index.js');

describe('arXiv search tests', () => {
	beforeEach(() => {
		mockAxiosGet.mockClear();
	});
	it('should return results as expected - default values', async () => {
		const results = await search({
			searchQueryParams: [
				{
					include: [{name: 'RNN'}, {name: 'Deep learning'}],
					exclude: [{name: 'LSTM'}],
				},
				{
					include: [{name: 'GAN'}],
				},
			],
		});
		expect(mockAxiosGet).toHaveBeenCalledWith(
			'http://export.arxiv.org/api/query?search_query=all:RNN+AND+all:Deep learning+ANDNOT+all:LSTM+OR+all:GAN&start=0&max_results=10'
		);
		expect(results).toMatchSnapshot();
	});
	it('should return results as expected', async () => {
		const results = await search({
			searchQueryParams: [
				{
					include: [{name: 'RNN'}, {name: 'Deep learning'}],
					exclude: [{name: 'LSTM'}],
				},
				{
					include: [{name: 'GAN', prefix: PREFIXES.CAT}],
				},
			],
			start: 10,
			maxResults: 50,
		});
		expect(mockAxiosGet).toHaveBeenCalledWith(
			'http://export.arxiv.org/api/query?search_query=all:RNN+AND+all:Deep learning+ANDNOT+all:LSTM+OR+cat:GAN&start=10&max_results=50'
		);
		expect(results).toMatchSnapshot();
	});
	it('should return results as expected - with sortBy and sortOrder', async () => {
		const results = await search({
			searchQueryParams: [
				{
					include: [{name: 'RNN'}, {name: 'Deep learning'}],
					exclude: [{name: 'LSTM'}],
				},
				{
					include: [{name: 'GAN'}],
				},
			],
			start: 10,
			maxResults: 50,
			sortBy: SORT_BY.RELEVANCE,
			sortOrder: SORT_ORDER.ASCENDING,
		});
		expect(mockAxiosGet).toHaveBeenCalledWith(
			'http://export.arxiv.org/api/query?search_query=all:RNN+AND+all:Deep learning+ANDNOT+all:LSTM+OR+all:GAN&start=10&max_results=50&sortBy=relevance&sortOrder=ascending'
		);
		expect(results).toMatchSnapshot();
	});
	it('should return results with meta data as expected', async () => {
		const results = await searchWithMeta({
			searchQueryParams: [
				{
					include: [{name: 'GAN'}],
				},
			],
			start: 10,
			maxResults: 50,
		});
		expect(mockAxiosGet).toHaveBeenCalledWith(
			'http://export.arxiv.org/api/query?search_query=all:GAN&start=10&max_results=50'
		);
		expect(results).toMatchSnapshot();
	});
	it('should export response parser', () => {
		expect(parseResponseData).toBeDefined();
		expect(typeof parseResponseData).toBe('function');
	});
	it('should use response parser', async () => {
		const src = require('./index');
		const original = src.parseResponseData;
		const spy = jest.fn(original);
		src.parseResponseData = spy;
		try {
			await src.search({
				searchQueryParams: [
					{
						include: [{name: 'GAN'}],
					},
				],
			});
			expect(spy).toHaveBeenCalledWith('XML');
		} finally {
			src.parseResponseData = original;
		}
	});
	it('should throw error - unsupported sortBy', async () => {
		await expect(
			search({
				searchQueryParams: [
					{
						include: [{name: 'GAN'}],
					},
				],
				sortBy: 'SORT_BY',
			})
		).rejects.toMatchSnapshot();
	});
	it('should throw error - unsupported sortOrder', async () => {
		await expect(
			search({
				searchQueryParams: [
					{
						include: [{name: 'GAN'}],
					},
				],
				sortOrder: 'SORT_ORDER',
			})
		).rejects.toMatchSnapshot();
	});
	it('should throw error - searchQueryParams is not an array', async () => {
		await expect(
			search({
				searchQueryParams: 'PARAMS',
			})
		).rejects.toMatchSnapshot();
	});
	it('should throw error - include tags is not an array', async () => {
		await expect(
			search({
				searchQueryParams: [
					{
						include: 'SOME_VALUE',
					},
				],
			})
		).rejects.toMatchSnapshot();
	});
	it('should throw error - exclude tags is not an array', async () => {
		await expect(
			search({
				searchQueryParams: [
					{
						include: [{name: 'GAN'}],
						exclude: 'SOME_VALUE',
					},
				],
			})
		).rejects.toMatchSnapshot();
	});
	it('should throw error - include tags empty', async () => {
		await expect(
			search({
				searchQueryParams: [
					{
						include: [],
					},
				],
			})
		).rejects.toMatchSnapshot();
	});
	it('should throw error - tag name is empty', async () => {
		await expect(
			search({
				searchQueryParams: [
					{
						include: [{name: ''}],
					},
				],
			})
		).rejects.toMatchSnapshot();
	});
	it('should throw error - tag name is not string', async () => {
		await expect(
			search({
				searchQueryParams: [
					{
						include: [{name: 123}],
					},
				],
			})
		).rejects.toMatchSnapshot();
	});
	it('should throw error - unsupported tag prefix', async () => {
		await expect(
			search({
				searchQueryParams: [
					{
						include: [{name: 'GAN', prefix: 'pre'}],
					},
				],
			})
		).rejects.toMatchSnapshot();
	});
});
