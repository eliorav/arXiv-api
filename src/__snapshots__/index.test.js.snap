// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`arXiv search tests should return results as expected - default values 1`] = `
Array [
  Object {
    "authors": Array [
      Array [
        "AUTHOR1",
      ],
      Array [
        "AUTHOR2",
      ],
    ],
    "categories": Array [
      "CATEGORY",
    ],
    "id": "PAPER_ID",
    "links": Array [
      Object {
        "href": "URL",
        "rel": "REL",
        "title": "TITLE",
        "type": "TYPE",
      },
    ],
    "published": "PUBLISHED",
    "summary": "SUMMARY",
    "title": "TITLE",
    "updated": "UPDATED",
  },
]
`;

exports[`arXiv search tests should return results as expected - with sortBy and sortOrder 1`] = `
Array [
  Object {
    "authors": Array [
      Array [
        "AUTHOR1",
      ],
      Array [
        "AUTHOR2",
      ],
    ],
    "categories": Array [
      "CATEGORY",
    ],
    "id": "PAPER_ID",
    "links": Array [
      Object {
        "href": "URL",
        "rel": "REL",
        "title": "TITLE",
        "type": "TYPE",
      },
    ],
    "published": "PUBLISHED",
    "summary": "SUMMARY",
    "title": "TITLE",
    "updated": "UPDATED",
  },
]
`;

exports[`arXiv search tests should return results as expected 1`] = `
Array [
  Object {
    "authors": Array [
      Array [
        "AUTHOR1",
      ],
      Array [
        "AUTHOR2",
      ],
    ],
    "categories": Array [
      "CATEGORY",
    ],
    "id": "PAPER_ID",
    "links": Array [
      Object {
        "href": "URL",
        "rel": "REL",
        "title": "TITLE",
        "type": "TYPE",
      },
    ],
    "published": "PUBLISHED",
    "summary": "SUMMARY",
    "title": "TITLE",
    "updated": "UPDATED",
  },
]
`;

exports[`arXiv search tests should return results with meta data as expected 1`] = `
Object {
  "entries": Array [
    Object {
      "authors": Array [
        Array [
          "AUTHOR1",
        ],
        Array [
          "AUTHOR2",
        ],
      ],
      "categories": Array [
        "CATEGORY",
      ],
      "id": "PAPER_ID",
      "links": Array [
        Object {
          "href": "URL",
          "rel": "REL",
          "title": "TITLE",
          "type": "TYPE",
        },
      ],
      "published": "PUBLISHED",
      "summary": "SUMMARY",
      "title": "TITLE",
      "updated": "UPDATED",
    },
  ],
  "itemsPerPage": 1,
  "link": "LINK",
  "startIndex": 0,
  "totalResults": 2,
  "updated": "2021-05-16T00:00:00-04:00",
}
`;

exports[`arXiv search tests should throw error - exclude tags is not an array 1`] = `[Error: include and exclude must be arrays]`;

exports[`arXiv search tests should throw error - include tags empty 1`] = `[Error: include is a mandatory field]`;

exports[`arXiv search tests should throw error - include tags is not an array 1`] = `[Error: include and exclude must be arrays]`;

exports[`arXiv search tests should throw error - searchQueryParams is not an array 1`] = `[Error: query param must be an array]`;

exports[`arXiv search tests should throw error - tag name is empty 1`] = `[Error: you must specify tag name]`;

exports[`arXiv search tests should throw error - tag name is not string 1`] = `[Error: you must specify tag name]`;

exports[`arXiv search tests should throw error - unsupported sortBy 1`] = `[Error: unsupported sort by option. should be one of: relevance lastUpdatedDate submittedDate]`;

exports[`arXiv search tests should throw error - unsupported sortOrder 1`] = `[Error: unsupported sort order option. should be one of: ascending descending]`;

exports[`arXiv search tests should throw error - unsupported tag prefix 1`] = `[Error: unsupported prefix: pre]`;
